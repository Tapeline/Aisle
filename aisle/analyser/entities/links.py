from dataclasses import dataclass

from aisle.parser.nodes.links import LinkType


@dataclass
class Link:
    type: LinkType
    to: str
    over: str | None
    description: str
