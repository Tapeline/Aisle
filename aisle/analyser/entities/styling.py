from dataclasses import dataclass

from aisle.parser.nodes.legend import LegendSelector


@dataclass(init=False)
class StylingAttributes:
    bg: str | None = None
    fg: str | None = None

    def __init__(self, **kwargs):
        for kw_k, kw_v in kwargs.items():
            setattr(self, kw_k, kw_v)


@dataclass
class LegendStyling:
    selector: LegendSelector
    attrs: StylingAttributes
    description: str
