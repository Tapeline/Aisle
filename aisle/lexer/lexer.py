"""Contains lexer implementation."""

from collections.abc import Sequence

from aisle.lexer.exceptions import (
    UnexpectedCharacterException,
    StringNotClosedException,
    IncompleteUnicodeQuadException,
)
from aisle.lexer.tokens import (
    Token,
    TokenType,
    RESERVED_STRINGS,
    KEYWORDS,
)


class Lexer:
    def __init__(
            self,
            source: str,
            tab_size: int = 4,
    ) -> None:
        self._src = _norm_newlines(source)
        self._lines = self._src.splitlines()
        self._i = 0
        self._start = 0
        self._line = 1
        self._line_start = 0
        self._tokens = []
        self._tab_size = tab_size

    @property
    def _at_end(self):
        """Check if reached end of source."""
        return self._i >= len(self._src)

    def _next(self) -> str | None:
        if self._i >= len(self._src):
            return None
        self._i += 1
        return self._src[self._i - 1]

    @property
    def _current(self) -> str | None:
        return None if self._at_end else self._src[self._i]

    def _add_token(
            self,
            token_type: TokenType,
            lexeme: str | None = None
    ) -> None:
        token = Token(
            token_type,
            lexeme or self._src[self._start:self._i],
            self._line
        )
        self._start = self._i
        self._tokens.append(token)

    def _match(self, pattern: str) -> str | None:
        if self._at_end:
            return None
        src = self._src[self._i:]
        if src.startswith(pattern):
            self._i += len(pattern)
            return pattern
        return None

    def _error(self, exc_type, *args, **kwargs) -> None:
        raise exc_type(
            *args,
            **kwargs,
            source=self._src,
            line=self._line
        )

    def scan(self) -> Sequence[Token]:
        while not self._at_end:
            self._scan_token()
        return self._tokens

    def _scan_token(self) -> None:
        c = self._next()
        match c:
            case "=":
                self._add_token(TokenType.ASSIGN)
            case ":":
                self._add_token(TokenType.COLON)
            case "(":
                self._add_token(TokenType.LPAR)
            case ")":
                self._add_token(TokenType.RPAR)
            case "[":
                self._add_token(TokenType.LBRACKET)
            case "]":
                self._add_token(TokenType.RBRACKET)
            case "-":
                if self._match("->"):
                    self._add_token(TokenType.ARROW_R)
                elif self._match("--"):
                    self._add_token(TokenType.ARROW_NO_DIR)
                else:
                    self._error(UnexpectedCharacterException, self._current)
            case "<":
                if self._match("->"):
                    self._add_token(TokenType.ARROW_BI_DIR)
                elif self._match("--"):
                    self._add_token(TokenType.ARROW_L)
                else:
                    self._error(UnexpectedCharacterException, self._current)
            case "\n":
                self._add_token(TokenType.NEWLINE)
                self._line += 1
                self._line_start = self._i
            case " ":
                if self._match(" " * (self._tab_size - 1)):
                    self._add_token(TokenType.INDENT)
                else:
                    self._start = self._i
            case "\t":
                self._add_token(TokenType.INDENT)
            case '"':
                self._scan_string()
            case _:
                self._i -= 1
                self._scan_text()

    def _scan_string(self) -> None:
        text = ""
        while not self._at_end and self._current != '"':
            if self._current == "\\":
                text += self._scan_escape_seq()
            else:
                text += self._current
                self._i += 1
        if not (not self._at_end and self._current == '"'):
            self._error(StringNotClosedException)
        self._i += 1
        self._add_token(TokenType.TEXT, text)

    def _scan_escape_seq(self) -> str:
        self._next()
        control = self._next()
        match control:
            case "r":
                return "\r"
            case "f":
                return "\f"
            case "n":
                return "\n"
            case "t":
                return "\t"
            case '"':
                return '"'
            case "\\":
                return "\\"
            case "u":
                code = int(self._scan_unicode_quad(), 16)
                return chr(code)
        self._error(UnexpectedCharacterException, char=control)

    def _scan_unicode_quad(self) -> str:
        code = ""
        for _ in range(4):
            c = self._next()
            if c is None:
                self._error(IncompleteUnicodeQuadException)
            if c not in "0123456789abcdefABCDEF":
                self._error(IncompleteUnicodeQuadException)
            code += c
        return code

    def _scan_text(self) -> None:
        subindent_level = self._i - self._line_start
        text = ""
        while not self._at_end:
            if self._current == "\\":
                text += self._scan_escape_seq()
            elif self._current == "\n":
                self._line += 1
                if not self._next_line_on_same_level(subindent_level):
                    break
                text += self._current
                text += " "*subindent_level  # Retain all indentation
                self._i += subindent_level + 1
            else:
                text += self._current
                self._i += 1
            if _get_reserved_string_at_end(text):
                break
        reserved = _get_reserved_string_at_end(text)
        if reserved and len(reserved) != len(text):
            self._i -= len(reserved)
            text = text.removesuffix(reserved)

        # Return all extra indentation and newline symbols back
        stripped_text = text.strip()
        suffix_start = text.index(stripped_text) + len(stripped_text)
        suffix_length = len(text) - suffix_start
        self._i -= suffix_length
        # Update line count accordingly
        self._line -= text[suffix_start:].count("\n")
        # Remove unnecessary indentation
        text = _remove_indentation(text, subindent_level)

        if text in KEYWORDS:
            self._add_token(TokenType.KEYWORD, text.strip())
        else:
            self._add_token(TokenType.TEXT, text.strip())

    def _next_line_on_same_level(self, subindent_level: int) -> bool:
        return self._lines[self._line - 1].startswith(" " * subindent_level)


def _get_reserved_string_at_end(text: str) -> str | None:
    if text in RESERVED_STRINGS:
        return text
    for string in RESERVED_STRINGS:
        if _should_be_split(text, string):
            return string
    return None


def _should_be_split(string: str, reserved: str) -> bool:
    if not string.endswith(reserved):
        return False
    trunk = string.removesuffix(reserved)
    if not reserved.isalnum():
        return True
    return trunk[-1] in "\n\r\t "


def _norm_newlines(text: str) -> str:
    """Replace all CRLF with LF."""
    return text.replace("\r\n", "\n").replace("\r", "\n")


def _remove_indentation(text: str, amount: int) -> str:
    lines = text.splitlines()
    prefix = " "*amount
    for i, line in enumerate(lines):
        lines[i] = line.removeprefix(prefix)
    return "\n".join(lines)

"""
tech = fdffd

"""