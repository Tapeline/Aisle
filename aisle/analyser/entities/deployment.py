from dataclasses import dataclass

from aisle.analyser.entities.project import ProjectEntity


@dataclass
class ServiceDeployment:
    service_name: str
    deploy_as: str


@dataclass
class DeploymentEntity(ProjectEntity):
    name: str
    description: str
    deploys: list[ServiceDeployment]
    tags: list[str]
