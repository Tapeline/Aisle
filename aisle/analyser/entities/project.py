from dataclasses import dataclass
from typing import Protocol

from aisle.analyser.entities.styling import LegendStyling


class ProjectEntity(Protocol):
    ...


@dataclass
class Project:
    name: str
    description: str
    namespace: dict[str, ProjectEntity]
    styling: list[LegendStyling]
    comments: list[str]
