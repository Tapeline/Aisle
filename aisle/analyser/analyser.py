from collections.abc import Sequence
from dataclasses import dataclass

from aisle.analyser.entities.containers import ServiceEntity
from aisle.analyser.entities.context import ActorEntity, SystemEntity
from aisle.analyser.entities.deployment import DeploymentEntity, ServiceDeployment
from aisle.analyser.entities.links import Link
from aisle.analyser.entities.project import Project
from aisle.analyser.entities.styling import LegendStyling, StylingAttributes
from aisle.analyser.exceptions import *
from aisle.parser.nodes.attribute import AttrNode
from aisle.parser.nodes.base import Node
from aisle.parser.nodes.entity import ProjectDefNode, EntityNode, EntityType, DeployNode
from aisle.parser.nodes.legend import LegendDeclarationNode
from aisle.parser.nodes.links import LinkCollectionNode
from aisle.parser.nodes.scope import ScopeNode, ScopeType
from aisle.parser.nodes.text import TextNode


class AnalyserBase:
    def _visit(self, node: Node):
        class_name = node.__class__.__name__
        self._pre_visit(node)
        try:
            getattr(self, f"_visit_{class_name}")(node)
        except AttributeError as error:
            raise VisitMethodNotFoundError(class_name, self) from error

    def _pre_visit(self, node: Node):
        ...


@dataclass
class Scope:
    type: ScopeType
    name: str


class Analyser(AnalyserBase):
    def __init__(self, nodes: Sequence[Node]):
        self.project: Project | None = None
        self.scope: Scope | None = None
        self.nodes = nodes

    def analyse(self) -> Project:
        for node in self.nodes:
            self._visit(node)
        if self.project is None:
            raise NoProjectDefinedException(Node(0), message="No project found")
        return self.project

    def _pre_visit(self, node: Node):
        if self.project is None and not isinstance(node, ProjectDefNode):
            raise NoProjectDefinedException(node)

    def _visit_ProjectDefNode(self, node: ProjectDefNode):
        if self.project is not None:
            raise DuplicateProjectDefinitionException(node)
        description = "\n".join(node.description)
        self.project = Project(
            name=node.name,
            description=description,
            namespace={},
            styling=[],
            comments=[]
        )

    def _visit_ScopeNode(self, node: ScopeNode):
        if self.project.name != node.scope_name:
            raise UnmatchedProjectAndScopeNameException(
                node,
                scope_name=node.scope_name,
                project_name=self.project.name
            )
        self.scope = Scope(node.scope_type, node.scope_name)

    def _ensure_entity_applicable(self, entity: EntityNode):
        if self.scope is None:
            raise UnmatchedScopeAndEntityTypeException(
                entity,
                scope_type=None,
                entity_type=entity.type.value
            )
        s_type = self.scope.type
        e_type = entity.type
        if not (
            s_type == ScopeType.CONTEXT and e_type == EntityType.ACTOR
            or s_type == ScopeType.CONTEXT and e_type == EntityType.SYSTEM
            or s_type == ScopeType.CONTAINERS and e_type == EntityType.SERVICE
            or s_type == ScopeType.DEPLOYMENT and e_type == EntityType.DEPLOYMENT
        ):
            raise UnmatchedScopeAndEntityTypeException(
                entity,
                scope_type=s_type.value,
                entity_type=e_type.value
            )

    def _visit_EntityNode(self, node: EntityNode):
        self._ensure_entity_applicable(node)
        match node.type:
            case EntityType.ACTOR:
                self._create_actor(node)
            case EntityType.SYSTEM:
                self._create_system(node)
            case EntityType.SERVICE:
                self._create_service(node)
            case EntityType.DEPLOYMENT:
                self._create_deployment(node)

    def _evaluate_entity_body(
            self, body: Sequence[Node]
    ) -> tuple[str, list[Link], dict[str, str]]:
        description = "\n".join(
            node.text for node in body if isinstance(node, TextNode)
        )
        attrs = {
            node.name: node.value
            for node in body if isinstance(node, AttrNode)
        }
        links = []
        for node in body:
            if isinstance(node, LinkCollectionNode):
                links.extend(node.links)
        links = [
            Link(
                type=link.type,
                to=link.to,
                over=link.over,
                description="\n".join(link.description)
            )
            for link in links
        ]
        return description, links, attrs

    def _create_actor(self, node: EntityNode):
        description, links, _ = self._evaluate_entity_body(node.body)
        entity = ActorEntity(
            name=node.name,
            description=description,
            links=links,
            tags=list(node.tags)
        )
        self.project.namespace[entity.name] = entity

    def _create_system(self, node: EntityNode):
        description, links, _ = self._evaluate_entity_body(node.body)
        entity = SystemEntity(
            name=node.name,
            description=description,
            links=links,
            tags=list(node.tags)
        )
        self.project.namespace[entity.name] = entity

    def _create_service(self, node: EntityNode):
        description, links, attrs = self._evaluate_entity_body(node.body)
        entity = ServiceEntity(
            name=node.name,
            description=description,
            links=links,
            tech=attrs.get("tech"),
            system=attrs.get("system"),
            tags=list(node.tags)
        )
        self.project.namespace[entity.name] = entity

    def _create_deployment(self, node: EntityNode):
        deploys = [
            ServiceDeployment(deploy.target, deploy.deploy_as)
            for deploy in node.body if isinstance(deploy, DeployNode)
        ]
        description, _, _ = self._evaluate_entity_body(node.body)
        entity = DeploymentEntity(
            name=node.name,
            description=description,
            deploys=deploys,
            tags=list(node.tags),
        )
        self.project.namespace[entity.name] = entity

    def _visit_TextNode(self, node: TextNode):
        self.project.comments.append(node.text)

    def _visit_LegendDeclarationNode(self, node: LegendDeclarationNode):
        description, _, attrs = self._evaluate_entity_body(node.body)
        style = LegendStyling(
            selector=node.selector,
            description=description,
            attrs=StylingAttributes(**attrs)
        )
        self.project.styling.append(style)
