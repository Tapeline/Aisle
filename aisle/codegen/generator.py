import hashlib
import shlex
import string
import textwrap
from typing import Final

from aisle.analyser.entities.context import ActorEntity, SystemEntity
from aisle.analyser.entities.containers import ServiceEntity
from aisle.analyser.entities.deployment import DeploymentEntity
from aisle.analyser.entities.links import Link
from aisle.analyser.entities.project import Project, ProjectEntity
from aisle.parser.nodes.links import LinkType


class SafeNameStorage:
    def __init__(self):
        self._names: dict[str, str] = {}

    def __getitem__(self, item) -> str:
        if item not in self._names:
            self._names[item] = _safe_name(item)
        return self._names[item]


class CodeGenerator:
    def __init__(self, project: Project):
        self.project = project
        self._safe_names = SafeNameStorage()

    def gen_actor(self, actor: ActorEntity) -> str:
        safe_name = self._safe_names[actor.name]
        tags = _compile_tags(actor.tags)
        return (
            f'Person('
            f'{safe_name}, '
            f'"{_safe_str(actor.name)}", '
            f'"{_safe_str(actor.description)}",'
            f'$tags="{tags}"'
            f')'
        )

    def gen_system(self, system: SystemEntity) -> str:
        safe_name = self._safe_names[system.name]
        tags = _compile_tags(system.tags)
        suffix = "_Ext" if system.is_external else ""
        return (
            f'System{suffix}('
            f'{safe_name}, '
            f'"{_safe_str(system.name)}", '
            f'"{_safe_str(system.description)}",'
            f'$tags="{tags}"'
            f')'
        )

    def gen_service(self, service: ServiceEntity) -> str:
        safe_name = self._safe_names[service.name]
        tags = _compile_tags(service.tags)
        suffix = "_Ext" if service.is_external else ""
        return (
            f'Container{suffix}('
            f'{safe_name}, '
            f'"{_safe_str(service.name)}", '
            f'"{_safe_str(service.description)}",'
            f'$tags="{tags}"'
            f')'
        )

    def gen_system_internals(self, system: SystemEntity) -> str:
        services = self.project.get_services_of_system(system.name)
        service_gen = "\n".join(map(self.gen_service, services))
        safe_name = self._safe_names[system.name]
        boundary = (
            f'System_Boundary('
            f'{safe_name},'
            f'"{_safe_str(system.name)}"'
            f')'
        )
        return boundary + "{\n" + _indent(service_gen, 4) + "\n}"

    def gen_link(self, link_from: ProjectEntity, link: Link) -> str:
        safe_name_a = self._safe_names[link_from.name]
        safe_name_b = self._safe_names[link.to]
        link_str = ""
        if link.type == LinkType.OUTGOING:
            link_str = (
                f'Rel('
                f'{safe_name_a}, '
                f'{safe_name_b}, '
                f'"{_safe_str(link.description)}",'
                f'"{_safe_str(link.over)}"'
                f')'
            )
        if link.type == LinkType.INCOMING:
            link_str = (
                f'Rel('
                f'{safe_name_b}, '
                f'{safe_name_a}, '
                f'"{_safe_str(link.description)}",'
                f'"{_safe_str(link.over)}"'
                f')'
            )
        if (
                link.type == LinkType.BIDIRECTIONAL
                or link.type == LinkType.NON_DIRECTED
        ):
            link_str = (
                f'BiRel('
                f'{safe_name_b}, '
                f'{safe_name_a}, '
                f'"{_safe_str(link.description)}",'
                f'"{_safe_str(link.over)}"'
                f')'
            )
        return link_str

    def gen_deployment(self, deployment: DeploymentEntity) -> str:
        code = []
        code += map(self.gen_deployment, deployment.inner_entities)
        for svc_deployment in deployment.deploys:
            code.append(
                f'Node('
                f'{self._safe_names[svc_deployment.service_name]},'
                f'"{_safe_str(svc_deployment.service_name)}",'
                f'$descr="{_safe_str(svc_deployment.deploy_as)}"'
                f')'
            )
        return (
            f'Boundary('
            f'{self._safe_names[deployment.name]},'
            f'"{_safe_str(deployment.name)}",'
            f'$descr="{_safe_str(deployment.description)}"'
            f')'
            + "{\n"
            + _indent("\n".join(code), 4)
            + "\n}"
        )

    def gen_context_map(self) -> str:
        code = []
        relations = []
        for actor in self.project.get_actors():
            code.append(self.gen_actor(actor))
            for rel in actor.links:
                relations.append((actor, rel))
        for system in self.project.get_systems():
            code.append(self.gen_system(system))
            for rel in system.links:
                relations.append((system, rel))
        for entity, rel in relations:
            code.append(self.gen_link(entity, rel))
        return "\n\n".join(code)

    def gen_container_map(self) -> str:
        imported_actors = []
        code = []
        relations = []
        used_entities = set()
        for system in self.project.get_systems():
            code.append(self.gen_system_internals(system))
            for service in self.project.get_services_of_system(system.name):
                used_entities.add(service.name)
                for rel in service.links:
                    relations.append((service, rel))
                    entity = self.project.namespace[rel.to]
                    if isinstance(entity, ActorEntity):
                        imported_actors.append(entity)
        for actor in imported_actors:
            code.append(self.gen_actor(actor))
        for entity, rel in relations:
            code.append(self.gen_link(entity, rel))
        return "\n\n".join(code)

    def gen_deployment_map(self) -> str:
        code = []
        for deployment in self.project.get_deployments():
            code.append(self.gen_deployment(deployment))
        return "\n\n".join(code)


_ALLOWED_CHARS: Final = string.ascii_letters + string.digits + "_"


def _safe_name(name: str) -> str:
    name = name.replace(" ", "_")
    filtered_name = "".join(char for char in name if char in _ALLOWED_CHARS)
    hash_part = hashlib.md5(name.encode()).hexdigest()[:8]
    if filtered_name == name:
        return filtered_name
    return f"{filtered_name}_{hash_part}"


def _indent(text: str, indent_level: int) -> str:
    return textwrap.indent(text, prefix=" " * indent_level)


def _compile_tags(tags: list[str]) -> str:
    return "+".join(map(_safe_name, tags))


def _safe_str(text: str) -> str:
    if text is None:
        return ""
    return text.replace('"', '\\"').replace("\n", "\\n")


def generate_context(project: Project) -> str:
    cg = CodeGenerator(project)
    return (
        f'@startuml\n'
        f'!include <C4/C4_Context>\n'
        f'{cg.gen_context_map()}\n'
        f'@enduml\n'
    )


def generate_containers(project: Project) -> str:
    cg = CodeGenerator(project)
    return (
        f'@startuml\n'
        f'!include <C4/C4_Container>\n'
        f'{cg.gen_container_map()}\n'
        f'@enduml\n'
    )


def generate_deployments(project: Project) -> str:
    cg = CodeGenerator(project)
    return (
        f'@startuml\n'
        f'!include <C4/C4_Deployment>\n'
        f'{cg.gen_deployment_map()}\n'
        f'@enduml\n'
    )
