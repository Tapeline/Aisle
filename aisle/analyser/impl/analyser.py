from collections.abc import Sequence
from dataclasses import dataclass
from types import MappingProxyType

from aisle.analyser.entities.containers import ServiceEntity
from aisle.analyser.entities.context import ActorEntity, SystemEntity
from aisle.analyser.entities.deployment import (
    DeploymentEntity,
    ServiceDeployment,
)
from aisle.analyser.entities.links import Link
from aisle.analyser.entities.project import Namespace, Project
from aisle.analyser.entities.styling import LegendStyling, StylingAttributes
from aisle.analyser.exceptions import (
    DuplicateProjectDefinitionError,
    NoProjectDefinedError,
    UnmatchedProjectAndScopeNameError,
    UnmatchedScopeAndEntityTypeError,
    VisitMethodNotFoundError,
)
from aisle.analyser.interfaces import AbstractAnalyser
from aisle.parser.nodes.attribute import AttrNode
from aisle.parser.nodes.base import Node
from aisle.parser.nodes.entity import (
    DeployNode,
    EntityNode,
    EntityType,
    ProjectDefNode,
)
from aisle.parser.nodes.legend import LegendDeclarationNode
from aisle.parser.nodes.links import LinkCollectionNode, LinkNode
from aisle.parser.nodes.scope import ScopeNode, ScopeType
from aisle.parser.nodes.text import TextNode


class _AnalyserBase:
    def _visit(self, node: Node) -> None:
        class_name = node.__class__.__name__
        self._pre_visit(node)
        try:
            getattr(self, f"_visit_{class_name}")(node)
        except AttributeError as error:  # pragma: no cover
            raise VisitMethodNotFoundError(class_name, self) from error

    def _pre_visit(self, node: Node) -> None:
        ...  # pragma: no cover


@dataclass
class Scope:
    """Analyser scope."""

    type: ScopeType
    name: str


_ALLOWED_ENTITY_TYPES = MappingProxyType(
    {
        ScopeType.CONTEXT: {EntityType.ACTOR, EntityType.SYSTEM},
        ScopeType.CONTAINERS: {EntityType.SERVICE},
        ScopeType.DEPLOYMENT: {EntityType.DEPLOYMENT},
    }
)


class Analyser(AbstractAnalyser, _AnalyserBase):
    """Default analyser impl."""

    def __init__(self, nodes: Sequence[Node]):
        """Create analyser."""
        self.project: Project | None = None
        self.scope: Scope | None = None
        self.nodes = nodes

    def analyse(self) -> Project:
        """Analyse AST and create project."""
        for node in self.nodes:
            self._visit(node)
        if self.project is None:
            raise NoProjectDefinedError(
                Node(0),
                message="No project found"
            )
        return self.project

    def _pre_visit(self, node: Node) -> None:
        if self.project is None and not isinstance(node, ProjectDefNode):
            raise NoProjectDefinedError(node)

    def _visit_ProjectDefNode(self, node: ProjectDefNode) -> None:
        if self.project is not None:
            raise DuplicateProjectDefinitionError(node)
        description = "\n".join(node.description)
        self.project = Project(
            name=node.name,
            description=description,
            namespace=Namespace(),
            styling=[],
            comments=[]
        )

    def _visit_ScopeNode(self, node: ScopeNode) -> None:
        if self.project.name != node.scope_name:
            raise UnmatchedProjectAndScopeNameError(
                node,
                scope_name=node.scope_name,
                project_name=self.project.name
            )
        self.scope = Scope(node.scope_type, node.scope_name)

    def _ensure_entity_applicable(  # pragma: no cover
            self,
            entity: EntityNode
    ) -> None:
        if self.scope is None:
            raise UnmatchedScopeAndEntityTypeError(
                entity,
                scope_type=None,
                entity_type=entity.type.value
            )
        if entity.type not in _ALLOWED_ENTITY_TYPES[self.scope.type]:
            raise UnmatchedScopeAndEntityTypeError(
                entity,
                scope_type=self.scope.type.value,
                entity_type=entity.type.value
            )

    def _visit_EntityNode(self, node: EntityNode) -> None:
        self._ensure_entity_applicable(node)
        match node.type:
            case EntityType.ACTOR:
                self._create_actor(node)
            case EntityType.SYSTEM:
                self._create_system(node)
            case EntityType.SERVICE:
                self._create_service(node)
            case EntityType.DEPLOYMENT:
                self._create_deployment(node)

    def _evaluate_entity_body(
            self, body: Sequence[Node]
    ) -> tuple[str, list[Link], dict[str, str]]:
        description = "\n".join(
            node.text for node in body if isinstance(node, TextNode)
        )
        attrs = {
            node.name: node.value
            for node in body if isinstance(node, AttrNode)
        }
        collected_links: list[LinkNode] = []
        for node in body:
            if isinstance(node, LinkCollectionNode):
                collected_links.extend(node.links)
        links = [
            Link(
                type=link.type,
                to=link.to,
                over=link.over,
                description="\n".join(link.description)
            )
            for link in collected_links
        ]
        return description, links, attrs

    def _evaluate_deployment_body(
            self,
            body: Sequence[Node]
    ) -> tuple[str, list[ServiceDeployment], list[DeploymentEntity]]:
        description = "\n".join(
            node.text for node in body if isinstance(node, TextNode)
        )
        svc_deployments = [
            ServiceDeployment(node.target, node.deploy_as)
            for node in body if isinstance(node, DeployNode)
        ]
        inner_deployments = []
        for node in body:
            if not (
                    isinstance(node, EntityNode)
                    and node.type == EntityType.DEPLOYMENT
            ):
                continue
            i_desc, i_svc, i_inner = self._evaluate_deployment_body(node.body)
            inner_deployments.append(
                DeploymentEntity(
                    name=node.name,
                    description=i_desc,
                    tags=list(node.tags),
                    is_external=node.is_external,
                    deploys=i_svc,
                    inner_entities=i_inner
                )
            )
        return description, svc_deployments, inner_deployments

    def _create_actor(self, node: EntityNode) -> None:
        description, links, _ = self._evaluate_entity_body(node.body)
        entity = ActorEntity(
            name=node.name,
            description=description,
            links=links,
            tags=list(node.tags)
        )
        self.project.namespace[entity.name] = entity

    def _create_system(self, node: EntityNode) -> None:
        description, links, _ = self._evaluate_entity_body(node.body)
        entity = SystemEntity(
            name=node.name,
            description=description,
            links=links,
            tags=list(node.tags),
            is_external=node.is_external
        )
        self.project.namespace[entity.name] = entity

    def _create_service(self, node: EntityNode) -> None:
        description, links, attrs = self._evaluate_entity_body(node.body)
        entity = ServiceEntity(
            name=node.name,
            description=description,
            links=links,
            tech=attrs.get("tech"),
            system=attrs.get("system"),
            tags=list(node.tags),
            is_external=node.is_external
        )
        self.project.namespace[entity.name] = entity

    def _create_deployment(self, node: EntityNode) -> None:
        description, svc_deploys, inner_entities = (
            self._evaluate_deployment_body(node.body)
        )
        entity = DeploymentEntity(
            name=node.name,
            description=description,
            deploys=svc_deploys,
            tags=list(node.tags),
            is_external=node.is_external,
            inner_entities=inner_entities
        )
        self.project.namespace[entity.name] = entity

    def _visit_TextNode(self, node: TextNode) -> None:
        self.project.comments.append(node.text)

    def _visit_LegendDeclarationNode(
            self,
            node: LegendDeclarationNode
    ) -> None:
        description, _, attrs = self._evaluate_entity_body(node.body)
        style = LegendStyling(
            selector=node.selector,
            description=description,
            attrs=StylingAttributes(**attrs)
        )
        self.project.styling.append(style)
