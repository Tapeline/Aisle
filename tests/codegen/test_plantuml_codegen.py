from pathlib import Path

from aisle.analyser.analyser import Analyser
from aisle.analyser.entities.project import Project
from aisle.codegen.generator import (
    generate_context,
    generate_containers,
    generate_deployments
)
from aisle.lexer.lexer import Lexer
from aisle.parser.parser import Parser


def _read_project(path: str) -> Project:
    src = Path(path).read_text()
    tokens = Lexer(src).scan()
    nodes = Parser(src, tokens).parse()
    project = Analyser(nodes).analyse()
    return project


def test_context_codegen(snapshot):
    """Test that context layer generates correctly."""
    project = _read_project("./tests/fixtures/test_1_a.aisle")
    assert generate_context(project) == snapshot


def test_containers_codegen(snapshot):
    """Test that containers layer generates correctly."""
    project = _read_project("./tests/fixtures/test_1_a.aisle")
    assert generate_containers(project) == snapshot


def test_deployment_codegen(snapshot):
    """Test that deployment layer generates correctly."""
    project = _read_project("./tests/fixtures/test_1_a.aisle")
    assert generate_deployments(project) == snapshot


def test_link_directions_codegen(snapshot):
    """Test that different link directions work correctly."""
    project = _read_project("./tests/fixtures/test_2.aisle")
    assert generate_context(project) == snapshot


def test_name_clashes_codegen(snapshot):
    """Test that clashing names are resolved correctly."""
    project = _read_project("./tests/fixtures/test_3.aisle")
    assert generate_context(project) == snapshot
