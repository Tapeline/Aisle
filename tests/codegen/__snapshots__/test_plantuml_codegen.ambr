# serializer version: 1
# name: test_containers_codegen
  '''
  @startuml
  !include <C4/C4_Container>
  
  
  AddElementTag("Yandex", $backgroundColor='#777700')
  
  
  
  
  
  System_Boundary(Blogging_System,"Blogging System"){
      Container(Frontend_Service, "Frontend Service", "UI",$tags="",$techn="Vite+Vue")
      Container(Backend_Service, "Backend Service", "Provides API\nMaybe split into microservices later",$tags="",$techn="Litestar, sqlalchemy")
      ContainerDb(DB_Service, "DB Service", "",$tags="database",$techn="Postgres")
  }
  
  System_Boundary(Metrics_System,"Metrics System"){
      Container_Ext(Metrics_Service, "Metrics Service", "",$tags="",$techn="None")
  }
  
  Person(User, "User", "Uses blogging app. Reads and creates posts",$tags="")
  
  Person(Admin, "Admin", "Blog moderator",$tags="")
  
  Rel(Frontend_Service, Backend_Service, "Send API requests","HTTP")
  
  Rel(User, Frontend_Service, "","")
  
  Rel(Admin, Frontend_Service, "","")
  
  Rel(Backend_Service, Metrics_Service, "Send metrics info on each request","HTTP")
  
  Rel(Backend_Service, DB_Service, "Stores info","")
  
  SHOW_LEGEND()
  @enduml
  
  '''
# ---
# name: test_context_codegen
  '''
  @startuml
  !include <C4/C4_Context>
  
  
  AddElementTag("Yandex", $backgroundColor='#777700')
  
  
  
  
  
  Person(User, "User", "Uses blogging app. Reads and creates posts",$tags="")
  
  Person(Admin, "Admin", "Blog moderator",$tags="")
  
  System(Blogging_System, "Blogging System", "Stores and manages blogs",$tags="")
  
  System_Ext(Metrics_System, "Metrics System", "Collects site metrics\nWe will use Yandex Metrics",$tags="")
  
  Rel(User, Blogging_System, "Uses","")
  
  Rel(Admin, Blogging_System, "Uses","")
  
  Rel(Admin, Metrics_System, "Reviews post metrics","")
  
  Rel(Blogging_System, Metrics_System, "","")
  
  SHOW_LEGEND()
  @enduml
  
  '''
# ---
# name: test_deployment_codegen
  '''
  @startuml
  !include <C4/C4_Deployment>
  
  
  AddElementTag("Yandex", $backgroundColor='#777700')
  
  
  
  
  
  Boundary(VPS_Deployment,"VPS Deployment",$descr="Docker containers with Docker compose"){
      Boundary(Docker_compose,"Docker compose",$descr=""){
          Node(Frontend_Service,"Frontend Service",$descr="Docker container")
          Node(Backend_Service,"Backend Service",$descr="Docker container")
          Node(DB_Service,"DB Service",$descr="Docker container")
      }
  }
  
  Boundary(Provided_Metrics,"Provided Metrics",$descr="Metrics Service is already provided"){
      Node(Metrics_Service,"Metrics Service",$descr="Cloud provider")
  }
  
  SHOW_LEGEND()
  @enduml
  
  '''
# ---
# name: test_link_directions_codegen
  '''
  @startuml
  !include <C4/C4_Context>
  
  
  Person(User, "User", "",$tags="")
  
  System(System, "System", "",$tags="")
  
  Rel(User, System, "","")
  
  Rel(System, User, "","")
  
  BiRel(System, User, "","")
  
  BiRel(System, User, "","")
  
  SHOW_LEGEND()
  @enduml
  
  '''
# ---
# name: test_name_clashes_codegen
  '''
  @startuml
  !include <C4/C4_Context>
  
  
  Person(Plus_c7da77da, "Plus+", "",$tags="")
  
  System(Plus, "Plus", "",$tags="")
  
  SHOW_LEGEND()
  @enduml
  
  '''
# ---
